Update CSS Best Practices (#408)
10up/Engineering-Best-Practices: _includes/markdown/CSS.md

Our websites are built mobile first, using performant CSS. Well-structured CSS yields maintainability and better collaboration which ultimately yields better client experiences.

> Below are our suggested best practices for writing CSS at Fueled.

Reviewer: We start by saying `At 10up...` while here we say at Fueled. We probably need leadership alignment here, but we should consolidate

Reviewer 2: @dmtrmrv @Antonio-Laguna 

Let's maintain consistency with the rest of the doc and only reference 10up - and not introduce Fueled at this time. 

Fueled branding will be more a branding overhaul conversation - if / when the powers that be determine that is a direction we want to go here. 

Good catch Antonio.

> Below are some of the best practices for writing CSS that we follow.

Reviewer 3: @dmtrmrv "Below are our suggested best practices for writing CSS at Fueled"

Author: Updated.

<h2 id="established-system" class="anchor-heading">Follow an Established System First {% include Util/link_anchor anchor="established-system" %}</h2>

There are countless ways to organize your CSS styles. From naming conventions to structuring files, there’s no single "right" way to do it. What’s important is understanding and respecting the system that’s already in place.

* **Assume intent**. Always operate under the assumption that existing patterns or conventions were created for a reason. Even if they seem unfamiliar or inefficient at first glance, dig deeper to understand the context.  
* **Learn before you suggest.** When you join a project, take the time to understand how the styles are organized and why they’re structured that way. Avoid making snap judgments or introducing changes before fully grasping the system.  

> * **Build on established patterns first.** Stick to the existing conventions, even if they feel unfamiliar unless there’s an obvious reason to deviate. Continuity is critical for maintainability and teamwork.  

Reviewer 3: @dmtrmrv this point kinda feels the same as the first point. Perhaps we should combine them

> * **When in doubt, align with technical leadership**. If you are unsure about patterns or systems to adhere to, follow the guidance and recommendation of a project's technical leadership (if viable). 

Reviewer: ```suggestion
* **When in doubt, align with technical leadership**. If you are unsure about patterns or systems to adhere to, follow the guidance and recommendations of a project's technical leadership (where feasible). 
```

* **When in doubt, align with technical leadership**. If you are unsure about patterns or systems to adhere to, follow the guidance and recommendation of a project's technical leadership (if viable). 

> Only propose changes if the current convention is genuinely problematic—for example, if it’s inconsistent, hard to scale, or causes errors. When you suggest a change, be ready to explain why the new approach is better and how it solves a specific problem.

Reviewer 3: `problematic, e.g "if it's inconsistent..."`

This approach helps avoid duplication—why have two things that do the same job? It’s inefficient and creates unnecessary confusion. Beyond that, following established patterns shows respect for your team’s work and fosters a healthy, collaborative relationship. Consistency in the codebase reflects a consistent and unified team.

<h2 id="document-system" class="anchor-heading">Establish And Document The System {% include Util/link_anchor anchor="document-system" %}</h2>

When you start a new codebase, pause before writing components and choose a CSS methodology—such as ITCSS, BEM, or a utility-first approach—that fits the project’s size, maintenance horizon, and team expertise. Committing to a structured methodology early establishes clear naming rules, predictable specificity, and sensible file organization. These guardrails ease onboarding, minimize refactors, and save both time and budget over the life of the project.

> If you opt to design a custom architecture, capture every key choice: folder layout, naming conventions, tooling, and any deviations from common patterns. A short explanation of why each decision was made helps future contributors understand the system quickly and avoids guesswork.

Reviewer 3: @dmtrmrv Im wondering if we should include something about AI here - certain AI prompts for specific frameworks that can help with context sharing or standardization. Even the use of AI to better understand the CSS implementation - where theres overlap / redundancy or bloated code.

Author: @dainemawer I think this is a great idea. I'm a bit worried that it gets too much into the weeds, and AI advice tends to get old quickly. I'll resolve this to move things forward, but feel free to add your thoughts here or in a separate PR if this is already merged by then.

> Keep this record where engineers will look first—typically in a /decisions/ directory inside the repository’s setup or onboarding documentation. Follow your organization’s standard documentation format so the notes remain consistent, searchable, and easy to maintain.

Reviewer 3: @dmtrmrv I really like this idea - is there a way that we can highlight it in a more meaningful / noticeable manner? Even if its bold and italicized for now - it would be great to follow this on with an internal blog post. Happy to help you write that if you are stretched for time. I think every project we've ever done needs this `folder` / convention.

Author: Made it bold for now, maybe we can create another block later!

<h2 id="content-agnostic" class="anchor-heading">Assume You Know Nothing About the Content {% include Util/link_anchor anchor="content-agnostic" %}</h2>

When building components, always start with the mindset that you have no control over the content they’ll handle. Even if the client or designer specifies certain limits, assume those constraints could change—or be ignored entirely.

> Content is unpredictable. You might not know how long a title will be, how many items a navigation menu will have, or the length of an excerpt in a card. Design decisions made today can quickly be overturned tomorrow when new requirements come in.

Reviewer 3: @dmtrmrv you may have outlined this further down, but do we have examples of CSS rules to use here? `line-clamp` etc?

Author: I don't want to mention specific rules here because it may seem like if you follow them, you are good. In practice, this is more about stress testing the components with edge cases.

> Prepare for edge cases. Your components must be resilient and robust enough to handle all reasonable variations, including unexpected ones. If you don’t account for these possibilities, it could lead to broken layouts, unreadable text, or poor user experiences.

Reviewer 3: Perhaps its worth linking out to articles by industry leaders here that speak about resilience, I know there have been a few major ones over the years like Defensive CSS

Author: Agree, added the link.

### Stress Test Your Components

When building a component, test it against extreme content scenarios. For example:

* A navigation menu with 20 items instead of 5\.  
* A card title that wraps onto three or more lines.  
* A card title that is just one word  
* An excerpt that’s 300 characters long—or empty.

> Building flexible, content-agnostic components makes your work more resilient to change. It reduces the need for constant adjustments and protects the layout from breaking when new content doesn’t fit the original plan. And let’s be real: the original plan often changes.

Reviewer: This is just my 2 cents but `Let's be real: the original plan often changes` feels conversational to me. I'd change it to something more formal for Best Practices. Something like this could also work.

> In practice, requirements often evolve throughout a project's lifecycle.

or 

> Since project requirements commonly evolve, building adaptable components prevents costly refactoring later.

Author: Agree updated

<h2 id="minimal-styles" class="anchor-heading">Write Only the Styles You Need, Don't Do Too Much {% include Util/link_anchor anchor="minimal-styles" %}</h2>

When writing CSS, stick to the styles you need and avoid overcomplicating. For example, centering a container horizontally often involves a common snippet:

```css
.container {

>   max-inline-size: 48rem;

Reviewer: This is something I don't agree with too much.  Mostly because this is transpiled by PostCSS Preset Env most of the time. This introduces cognitive load into the developer for a use case that I don't think we have ever seen (in which `inline-size` means something different than `width`)

Author: Made more about the concept, not logical properties.

  max-inline-size: 48rem;
  margin: 0 auto;
}
```

> While this centering approach achieves the goal, it also zeroes out the top and bottom margins, which may not be intentional and will have unwanted impact in the cascade. Instead, use logical properties:

Reviewer 3: Great suggestion / advice.

```css
.container {
  max-inline-size: 48rem;
  margin-inline: auto;
}
```

This approach, using logical properties like `margin-inline`, maintains the desired horizontal centering without unnecessary side effects. Avoid shorthands that set multiple properties when only one is needed.

Another example is background styling: 

```css
.button--secondary {
  background: var(--color--dark-gray);
}
```

> Using background shorthand like the above implicitly sets many background properties, which can unintentionally override defaults. Instead, specify only the needed properties:

Reviewer 3: @dmtrmrv we should make sure that this applies to other property short hands like `border` and `transform` / `transition`

Reviewer: Nice point! 

Yep, also `font`!

Author: Agree, added.

```css
.button--secondary {
  background-color: var(--color--dark-gray);
}
```

Finally, be cautious with broad declarations that affect more elements than intended. For instance, adding an underline animation to links:

```css
a {
  text-decoration: none;
  background-image: linear-gradient(currentColor, currentColor);
  background-position: 0% 100%;
  background-repeat: no-repeat;
  background-size: 100% 2px;
  transition: background-size 0.3s;
}
```

Adding a global style like this can unintentionally affect other links that don’t require this underline treatment (image links, site logo, etc.). Instead, there are two recommended methods to handle similar styles. First, you can target elements (i.e. links) like this more precisely:

```css

> .entry-content :is(h1, h2, h3, h4, h5, h6, p, li) > a {

Reviewer 3: @dmtrmrv worth noting the use of `:is` here and possibly a link out to MDN

Author: Added!

Always advocate for creating mobile designs if they are not already included and for simplicity in implementation.

### Keep an Eye Out for Common Pitfalls

When reviewing designs, look for common pitfalls like the ones below:

* Different aspect ratios of images on mobile vs. desktop. If approved, this leads to complicated back-end logic and art direction.  
* Different order of elements on desktop vs mobile. This can usually be done with CSS Grid or Flexbox, but there are accessibility implications like focusing order that should be taken into account. This can usually be avoided.  
* Designs that require different markup on mobile vs. desktop. This often leads to bloated CSS, unnecessary JS to handle markup changes, or duplicated markup. All of this can usually be avoided.

If you see a way to simplify the implementation, always voice it. Suggest adjustments that will make the system leaner and more maintainable.

### Style the Project as If Desktop Doesn’t Exist First

Style the project as if Desktop doesn’t exist first and then expand or override rules for the desktop.

> With the capabilities of modern CSS it is usually best to reach for media queries last and use features like auto-rows in CSS Grid or functions like clamp() to create layouts that are mobile and desktop-friendly.

Reviewer 3: @dmtrmrv and example of `auto` values in CSS grid or `minmax` here would be beneficial to the team.

Reviewer: While I agree, I'm not sure this should live within a Best Practices doc?

Author: Yeah, I'm not sure that giving a specific example here is required. It can create a false sense of "I follow these, I'm doing okay."

Allowing for this flexibility helps handle the “in-between-breakpoint” layouts nicely.

### Scrolling Is Okay

Just like books have pages and readers know they need to flip them to move forward, users know they need to scroll to get further on a webpage. Keep this in mind when you see designs like:

* Lists of links that look like a select dropdown on mobile and turn into a list on desktop.  
* Grids on desktop that turn into a carousel on mobile.

There are cases where such layout patterns are necessary. Not all select dropdowns or carousels are bad, but in most cases, this is unnecessary complication. Simple scrolling usually works just fine and is more intuitive than more complicated UI patterns.

Needless to say, implementing such changes between viewports takes more time and can pose accessibility implications.

**Useful resources:**

- [There Is No Fold](https://www.lukew.com/ff/entry.asp?1946)

<h2 id="solid-defaults" class="anchor-heading">Set Solid Defaults {% include Util/link_anchor anchor="solid-defaults" %}</h2>

> When building a robust and scalable front-end architecture, setting good default styles is an essential first step. Establishing a CSS foundation ensures consistency across browsers and provides a solid base for further styling. This foundation can be achieved using one of three primary approaches:

Reviewer 3: @dmtrmrv I feel like we should be more intentional here about what we advise - or at least illustrate an example - I know on my projects I use a combination of `reset` and `modern-normalize` - I think what you're outlining here is super helpful, but it will lead to a lot of questions.

Reviewer 2: I feel like this is something we can provide internal guidance for with more intent. For example, establish our recommended Hybrid approach within our scaffolding ... where this is more agnostic guidance / best practices beyond the internal walls/ specific projects & builds at Fueled. 

That being said, we _could_ provide an example of what a Hybrid approach might look like ... but it would also be something we'd need to maintain OR reference a 3rd party example/article as a reference point. 

@dainemawer @dmtrmrv

Author: I agree with the idea of providing more guidance, but I think it should live in the scaffold. Once we have it updated, we can add a link to it here that will serve as an example.

1. **CSS Reset:** A reset removes virtually all default styles applied by browsers, leveling the playing field by eliminating discrepancies. While this approach offers the cleanest slate, it also removes potentially useful styles, requiring more effort to rebuild a cohesive design system.  
2. **CSS Normalize:** A normalize CSS file (i.e. [normalize.css](https://necolas.github.io/normalize.css/)) focuses on addressing inconsistencies between browsers while retaining most of their default styles. This approach minimizes the need for excessive rework while still providing a reliable base.  
3. **Hybrid (or Custom) Approach:** Many projects adopt a blend of resets and normalization, coupled with project-specific global styles, to balance control and practicality. For example, resetting some elements like margins and paddings but setting global defaults for others, (following ITCSS architecture) such as lists and forms.

> ### The importance of Global Rules

Reviewer: I wanted to call out something minor, which is consistency with title casing. There's a mix of title case and sentence case.

This should probably be 

```suggestion
### The Importance of Global Rules
```

### The importance of Global Rules

> Global rules play a critical role in setting consistent behavior across your entire application. These rules can prevent costly refactoring and unexpected bugs as your project evolves. One key example is setting the `box-sizing` property globally:

Reviewer 3: @dmtrmrv specifically for `box-sizing` I think `normalize.css` does this for you? We should be specific about this considering its a universal rule, we dont want engineers duplicating it if its already included in css partials within the ITCSS architecture

Reviewer 2: Or, we can state ... "One key example, if not already included in your defaults, is setting the `box-sizing` property globally" ... perhaps, adding another example that is not included in `normalize.css` would also be helpful.

Author: Addressed this one!

```css
*,
*::before,
*::after {
  box-sizing: border-box;
}

:where(body) {
  margin: 0;
}
```

> By applying box-sizing: border-box, you ensure that padding and border widths are included in the element’s total width and height calculations. This eliminates common layout issues and simplifies sizing calculations, reducing developer frustration and the likelihood of design inconsistencies.

Reviewer 2: ```suggestion
By applying `box-sizing: border-box`, you ensure that padding and border widths are included in the element’s total width and height calculations. This eliminates common layout issues and simplifies sizing calculations, reducing developer frustration and the likelihood of design inconsistencies.
```

It is very easy to set up at the beginning of a project and very hard and expensive to fix when the project is live and has hundreds (or in some cases thousands) of pages and many components.

### Foundational Styles for HTML Elements

A well-crafted CSS foundation includes thoughtful defaults for HTML elements, such as headings, paragraphs, and lists. Avoid setting specific styles, like `text-align: center`, on elements that are likely to be used in various contexts. For example:

**Headings:** Define consistent font sizes, weights, and spacing for `<h1>` through `<h6>` elements to create a harmonious typography hierarchy that cascades effectively throughout your project.

```css

> :where(h1) {

Reviewer 3: @dmtrmrv lets link out to the use of `:where` here - its still relatively new and comes with spicy specificity impacts.

Author: I've added it where it was first mentioned.

  /* styles here */
}

:where(h2) {
  /* styles here */
}

/* And so on for other heading levels */
```

**Lists:** Ensure that unordered and ordered lists have adequate styling by default

```css
:where(ul, ol, dl) {
  /* styles here */
}
```

These are just a few base examples. Foundational styles should be applied to all elements used in the project. Typically, this is the “elements” level in the ITCSS methodology.

### Avoiding Overly Opinionated Defaults

While foundational styles are crucial, avoid overly opinionated rules that assume specific use cases for elements. For example, setting `text-align: center` or fixed widths on common elements like `<div>` or `<p>` can lead to unnecessary overrides and limited reusability. Foundational styles should support flexibility and adaptability, empowering developers to craft unique designs without battling restrictive defaults.

> By thoughtfully implementing and balancing resets, normalization, and custom global styles, you can establish a CSS foundation that fosters maintainability, consistency, and scalability for your project.

Reviewer 3: I often think that these "preset" or "base" layer CSS partials should be 100% standard on every project we build. Worth mentioning if they are already built into our frameworks and tooling

Reviewer 2: I like that suggestion, but also remember this best practices is to be somewhat agnostic - where we can be more prescriptive within our tooling or internal guides.

Author: Yeah, once we update the scaffold, we can add a link here.

<h2 id="inverted-triangle" class="anchor-heading">Follow Inverted Triangle Architecture {% include Util/link_anchor anchor="inverted-triangle" %}</h2>

> Inverted Triangle is a methodology for organizing CSS, starting from broad, global styles down to more specific, local rules.

Reviewer 3: @dmtrmrv lets link out to this, to Harry Roberts article / video on this methodology

Author: Agree, added!

Begin with base styles that affect the entire application:

### Global

> * font-face: Defines the fonts available throughout the application.

Reviewer: [nit]

```suggestion
* `@font-face`: Defines the fonts available throughout the application.
```

> * media queries: When using tools like PostCSS, @custom-media rules should be placed here for global breakpoints.

Reviewer: ```suggestion
* media queries: When using tools like PostCSS, `@custom-media` rules should be placed here for global breakpoints.
```

* media queries: When using tools like PostCSS, @custom-media rules should be placed here for global breakpoints.

* CSS Custom Properties: Also known as variables. When placed in :root, they provide access to all components across the application.

* mixins: If using PostCSS, global mixins should also be defined at this level.

* defaults: Set application-wide defaults, such as box-sizing, global outline behavior, and other foundational styles. Keep this section minimal, as there typically aren’t many styles applied at this level.

### Elements

This layer targets raw HTML elements (tags), without using classes, IDs, or other specific selectors. The purpose of the Base level is to establish foundational styles for elements to be used within components later on.

Below are some examples of styles at this level:

```css
:where(h1) {

>     color: var(--color--heading);

Reviewer 2: @dmtrmrv align indents here

  font-family: var(--font--family--heading);
  font-size: var(--font--size--heading-1);
  font-weight: var(--font--weight--heading);
  line-height: var(--font--line-height--heading-1);
  margin-block: 0;
}
```

Key details to note:

* `:where` pseudo-class: Ensures zero specificity for these styles, making it easier to apply other styles

* CSS Custom Properties: These variables, defined at the global level, keep the styles consistent and easy to maintain.

> * Margin reset: The margin is set to zero to remove default browser margins. You can handle multiple headings in one go with :where(h1, h2, h3...) { margin-block: 0 }.

Reviewer 2: ```suggestion
* Margin reset: The margin is set to zero to remove default browser margins. You can handle multiple headings in one go with `:where(h1, h2, h3...) { margin-block: 0 }`.
```

> Avoid adding styles that are too opinionated at the base level. For example, setting text-align: center on the heading in the previous example can cause issues. Even if most headings are center-aligned, there will likely be cases where you need a left- or right-aligned heading. This forces you to override the style in all components where center alignment isn’t needed, bloating the CSS and adding mental overhead.

Reviewer 2: ```suggestion
Avoid adding styles that are too opinionated at the base level. For example, setting `text-align: center` on the heading in the previous example can cause issues. Even if most headings are center-aligned, there will likely be cases where you need a left- or right-aligned heading. This forces you to override the style in all components where center alignment isn’t needed, bloating the CSS and adding mental overhead.
```

Be cautious about which styles you place at this level. Only include styles that should apply to all elements in 99.9% of cases. Removing base-level styles once a project is live—especially if thousands of pages depend on them—can introduce regressions.

### Components

The Component layer is where styles are applied to reusable UI elements. These are more complex than base-level styles and use classes to ensure component-level specificity.

#### Key Guidelines

* Reusability: Focus on creating modular, reusable components. Always consider how the component will respond to:

  * Changes in viewport width.

  * Varying content. For example, if you're styling a card with a title that fits nicely on one line in the design, think about how the card will react to a very long title, or if there’s no title at all.

* Inheritance from Base: Components should take advantage of the solid foundation provided by base styles. When base-level styles are set up correctly, you minimize the need to redefine styles. Usually, you’ll only need to adjust specific properties like colors or spacing at the component level.

> * Avoid Using Margins on Components: Components should be built in isolation, meaning we don’t know exactly where a component will be used. Including margins directly in component styles can create unnecessary complications. We'll cover alternative approaches later in this document.

Reviewer 3: @dmtrmrv - I think its worth suggesting the approach to actually using margins at this point - as in, should engineers create margin utility classes that can be added as and when needed?

Reviewer 2: Dmitry mentions the solution further down - to use the parent element to define component margins rather than applying them to the component themselves. 

We may want to make clear that the intent is not that aren't suggesting we don't use margins _within_ a component, but rather we are suggesting to avoid adding margins to the component itself and leverage a vertical rhythm established by the container element. 

Perhaps a little more clarity on the intent is warranted while we still "cover alternative approaches later in the document" 

To be clear, the intent is to avoid - not to say there can't be exceptions - as with any of these guidelines.

Author: Addressed this.

> * Encapsulation: Component styles should be scoped and encapsulated to avoid affecting other components. Use methodologies like BEM or CUBE to ensure your styles remain modular and isolated.

Reviewer 3: 👍🏻

* Leverage Variables: Use globally defined variables for properties like colors and spacing. This avoids magic numbers and ensures visual consistency across your app.

* Responsiveness: Components should be fully responsive, adapting effortlessly to different container sizes and screen widths.

### Utilities

Utility classes provide quick, reusable styling solutions for common, single-purpose tasks. They allow you to apply styles without the need for creating new components.

#### Key Guidelines

* Single Responsibility: Each utility class should only do one thing. For example, is-style-h1 should only apply styles required to make the text appear as heading level 1 and do nothing else.

> * Specificity Level: Utility classes sit at the top of inverted triangle architecture, so they can have higher specificity then base styles and components. In practice they usually have the same level of specificity as components, because of that they should be applied after the component class.

Reviewer: ```suggestion
* Specificity Level: Utility classes sit at the top of the inverted triangle architecture, so they can have higher specificity than base styles and components. In practice, they usually have the same level of specificity as components, because of that they should be applied after the component class.
```

* Consistency: Like global variables, utility classes help maintain consistency throughout the app. By standardizing common patterns like text alignment, padding, or display properties, you reduce the risk of divergent styles across components.

**Useful resources:**

- [The Inverted Triangle Architecture: how to manage large CSS Projects](https://www.freecodecamp.org/news/managing-large-s-css-projects-using-the-inverted-triangle-architecture-3c03e4b1e6df/)

> <h2 id="accessibility" class="anchor-heading">Always Keep Accessibility in Mind {% include Util/link_anchor anchor="accessibility" %}</h2>

Reviewer: Very strong section! Nice one!

I'm lacking a more authoritative guideline across Reduced Motion, which has been removed from the doc. This is what I always have in mind when doing them 

https://css-tricks.com/nuking-motion-with-prefers-reduced-motion/

Let me know if you want me to do some writing here!

Reviewer 2: There is also this for reference: https://motion.dev/blog/web-animation-performance-tier-list

Author: Hey @Antonio-Laguna, following up on this. Yes, please update this, this would be great!

<h2 id="accessibility" class="anchor-heading">Always Keep Accessibility in Mind {% include Util/link_anchor anchor="accessibility" %}</h2>

### Keyboard Navigation

Make sure all interactive elements are accessible via the keyboard. Test that users can navigate through the interface using the Tab key and activate actions with Enter or Space. Avoid creating elements that require a mouse or touch to work.

### Color Contrast

Ensure there’s enough contrast between text and its background. Use tools to check for WCAG compliance. Aim for a contrast ratio of at least 4.5:1 for normal text and 3:1 for larger text. Avoid relying on color alone to convey information.

### Focus

Properly manage focus states. Use :focus styles to make focused elements visible and distinct. Don’t trap focus in modal dialogs or dropdowns without providing a clear way to exit. Reset focus logically when users interact with dynamic content, such as moving to a new page.

### Zoom Level

> Style every component to make sure zoom level up to 400% is supported. If you are using a sticky header, make sure that at 400% it doesn’t cover the majority of the screen.

Reviewer 3: @dmtrmrv this can be tricky to handle and often an edge case - I think we'd need some more defined guidance on how best to achieve this.

Reviewer 2: For some clarity on this, from what I understand... where the 400% zoom comes into play to meet WCAG AA:

If a user is viewing a page on a 1280pixel wide viewport, they should be able to zoom 400% and achieve the 320px viewport view and have the site work and be viewed as expected. In these instances, the font size should always be able to achieve AT LEAST a 200% size enlargement... all the while the reflow of the page is not disrupted. 

We could link out to this: https://www.w3.org/WAI/WCAG22/Understanding/reflow.html

Author: Updated this one to be more in-depth and agnostic without focusing too much on the sticky header.

* Avoid placing text directly over images unless sufficient contrast or a background overlay is provided to ensure readability.  
* Use native HTML `<select>` elements for dropdowns instead of creating custom solutions or using frameworks. Native elements are inherently accessible and offer consistent behavior across platforms.  
* When custom components are necessary, replicate the behavior of native elements. For instance, custom modals should support full keyboard navigation and include options to close them using standard shortcuts.  
* Design forms with clear and descriptive error messages. Pair these messages with ARIA roles or properties to make them accessible to screen readers.

<h2 id="low-specificity" class="anchor-heading">Keep Specificity Extremely Low {% include Util/link_anchor anchor="low-specificity" %}</h2>

Maintaining low specificity in your CSS is essential for creating scalable, maintainable, and predictable stylesheets. High specificity can make it difficult to override styles, leading to a cascade that is hard to manage and debug. By keeping specificity low, you ensure that your styles remain flexible and easy to adapt as your project evolves.

> The maximum specificity level you should aim for is 0, 1, 0 to 0, 2, 1\. If your specificity exceeds 0, 3, 1, you are likely doing something wrong and should revisit your selectors.

Reviewer 3: @dmtrmrv - wonder if there are any tools / chrome extensions that can help engineers figure this out on the fly?

Reviewer 2: We _could_ add a stylelint rule for `selector-max-specificity` in our scaffolds. Maybe this is a discussion for one of the makers calls? 

```
{
  "selector-max-specificity": "0,2,1"
}
```

More here: https://stylelint.io/user-guide/rules/selector-max-specificity/

10up toolkit link: https://github.com/10up/10up-toolkit/blob/develop/packages/stylelint-config/README.md

Reviewer: You recommend a maximum of “0, 1, 0 to 0, 2, 1” and say “If your specificity exceeds 0, 3, 1…” That’s sensible! 

We may want to explicitly mention avoiding tag selectors combined with classes unless necessary to keep within those thresholds.

This could also be mentioned since it's been used a lot to check for specificity https://specificity.keegan.st/

### Use `:where()` for Bare Elements

> Utilize the :where() pseudo-class when styling bare elements to avoid increasing specificity. Unlike other selectors, :where() doesn’t contribute to specificity, making it easier to override styles when needed and ensuring a cleaner cascade hierarchy. This allows you to apply styles without affecting how CSS rules are prioritized and inherited, keeping the order of styles predictable and easier to manage. For example:

Reviewer 2: ```suggestion
Utilize the `:where()` pseudo-class when styling bare elements to avoid increasing specificity. Unlike other selectors, `:where()` doesn’t contribute to specificity, making it easier to override styles when needed and ensuring a cleaner cascade hierarchy. This allows you to apply styles without affecting how CSS rules are prioritized and inherited, keeping the order of styles predictable and easier to manage. For example:
```

```css
:where(h1, h2, h3, h4, h5, h6) {
  margin-block: 0;
}
```

### Avoid Excessive Nesting

Limit the depth of nesting in your CSS selectors. Two levels of nesting are acceptable, and three should only be used when absolutely necessary. Excessive nesting can make CSS harder to read and maintain. It also increases specificity unnecessarily, which complicates overriding styles and debugging. Over-nesting complicates maintenance and increases specificity unnecessarily.

It’s okay to nest pseudo classes that represent state, for example:

```css
/* block */
.button {

  /* ✅ state inside the block is fine */
  &:hover {
    /* styles here */
  }
}

/* ✅ element and modifier stay flat */
.button__icon {
  /* styles here */
}

.button--primary {
  /* styles here */
}
```

> Not necessary to nest elements or modifiers:

Reviewer: ```suggestion
It's not necessary to nest elements or modifiers:
```

Not necessary to nest elements or modifiers:

```css
/* block */
.button {

  /* state inside the block is fine */
  &:hover {
    /* styles here */
  }

  /* ❌ element and modifier must not be nested inside the block */
  & .button__icon {
    /* styles here */
  }

  & .button--primary {
    /* styles here */
  }
}
```

### Avoid IDs

> Refrain from using IDs for styling purposes. IDs have high specificity, which makes them difficult to override and maintain. Instead, use classes for consistent and reusable styling.

Reviewer 3: @dmtrmrv - one technique I use is attribute selectors: `[id="drawer-trigger"]` - this still gives you a unique rule without the specificity cost.

### Avoid \!important

> Reserve \!important for truly exceptional cases. Overusing \!important disrupts the cascade and makes debugging more difficult. Focus on writing clean, organized selectors instead of relying on this property to enforce styles.

Reviewer 2: ```suggestion
Reserve `\!important` for truly exceptional cases. Overusing `!important` disrupts the cascade and makes debugging more difficult. Focus on writing clean, organized selectors instead of relying on this property to enforce styles.
```

<h2 id="no-margins-components" class="anchor-heading">Do Not Use Margins on Reusable Components {% include Util/link_anchor anchor="no-margins-components" %}</h2>

Margins might look like one of the easiest CSS properties to manage, but they’re also a common source of layout headaches if not handled carefully.

* Avoid margins on reusable components. Margins can lock components into specific contexts, making them harder to use in different layouts. If you add margins to a reusable component, you’re creating assumptions about where and how it will be placed.  
* Design components with isolation in mind. Treat every component as if you have no idea what its future container will look like. Focus on the component's internal spacing and leave external spacing to the parent or container that uses it.

### Example

> Let’s say you have a Card component. Avoid doing this:

Reviewer 2: @dmtrmrv 

Now that I see this, I am wondering if we should add an example where we use relative values (i.e. `em`) to provide relative margin depending on context for vertical rhythm (i.e. headings)? I know you mention the lobotomized owl below, but this is an area I do see folks get hung up on - so a quick and easy example might be helpful. I 100% defer to you though! Simply a suggestion. If you feel this is an anti-pattern - please ignore.

Author: I've added this a bit further down

<h2 id="one-direction-margins" class="anchor-heading">Apply Margins in One Direction {% include Util/link_anchor anchor="one-direction-margins" %}</h2>

When applying margins, stick to using them in a single direction. This approach promotes consistency, simplifies debugging, and makes layouts easier to manage.

* Consistency across the codebase. By standardizing margin usage, you create a predictable structure. This consistency makes it easier for others (and future you) to understand and work with the code.  
* Simpler debugging. Troubleshooting layout issues becomes much easier when margins are applied in only one direction. You don’t have to chase down conflicting styles or figure out which element's margins are responsible for extra spacing.  
* Avoid collapsing margins. Margins applied in both directions can lead to confusing margin-collapsing behavior in CSS. Sticking to one direction eliminates this issue entirely, making the layout behavior more predictable and easier to control.  

> * Why "up" is preferred. Applying margins upwards (e.g., margin-top) is particularly effective because it works seamlessly with the lobotomized owl selector (`* + *`). This selector targets all adjacent siblings and is perfect for applying consistent spacing between elements without extra markup.

Reviewer 3: Lets link out to an article on why the lobotomized owl selector is a good approach

Reviewer: Like this one from Heydon Pickering (I think he came up with the term)

https://alistapart.com/article/axiomatic-css-and-lobotomized-owls/

Author: Agree, added.

<h2 id="responsiveness" class="anchor-heading">Responsiveness and Media Queries {% include Util/link_anchor anchor="responsiveness" %}</h2>

Creating responsive designs is essential, but the approach you take can make a big difference in your code's maintainability and flexibility. Media queries are a powerful tool, but they shouldn’t always be your first choice.

* **Reach for media queries last.** Instead of relying heavily on media queries to adjust styles, focus on building layouts and components that adapt naturally to different screen sizes. Media queries should be a fallback for cases where intrinsic solutions won’t work.  

> * **Leverage clamp().** Use the clamp() function to create fluid, responsive values for properties like font sizes, spacing, and widths. It allows you to define a minimum, preferred, and maximum value, so elements scale dynamically without needing media queries.  

Reviewer 3: I wrote an indepth article for SitePoint on Fluid Typography if you'd like to link out to it: https://www.sitepoint.com/fluid-typography-css-clamp-function/

Author: Agree, added.

> * **Use intrinsic layouts with CSS Grid.** Grid layouts are inherently responsive and let you define flexible areas that adjust automatically to available space. Pair grids with techniques like minmax() and auto-fit for dynamic behavior without extra breakpoints.

Reviewer 3: This is huge and an often poorly considered point - would love to expand on this further

Reviewer 2: Agreed - at least maybe a link to a 3rd party article to start. This would be a great Learning & Knowledge session internally!

Reviewer 2: @dmtrmrv 

[Web animation performance tier list](https://motion.dev/blog/web-animation-performance-tier-list)

Author: @dainemawer I agree that more info here would be ideal; however, I feel the next step is to produce internal guidance on intrinsic layouts, i.e., Notion post. I'd love some help with this if you are interested! Going to close this so we can move on.

<h2 id="naming" class="anchor-heading">Naming Things {% include Util/link_anchor anchor="naming" %}</h2>

Clear, consistent naming makes your CSS maintainable, readable, and scalable. Good naming conventions help developers quickly understand what each part of the code does, leading to fewer mistakes and easier updates. Here are practical recommendations for naming CSS elements:

### Classes

* Use descriptive, semantic names (e.g., .user-profile, .card-header).  
* Prefer lowercase with hyphens (kebab-case).  
* Keep names reusable and avoid overly specific context.

### IDs

* Use sparingly and not for styling, (e.g. for use with anchors and in-page links)  
* Names should clearly communicate the unique purpose.  

> * Prefer camelCase (e.g., `#mainNavigation`).

Reviewer 3: @dmtrmrv id's are often used for JS related functionality - I've often enforced: `js-main-navigation` so that we know that JS is leveraging `getElementById` on a particular element - wonder if its worth including a convention like that here

Author: Agree, added!

### Custom Properties (CSS Variables)

* Name according to usage rather than appearance (e.g., `--primary-color`, `--spacing-large`).  
* Use hyphenated lowercase consistently.  
* Keep them semantic to represent meaning and intent.

> A common use case for CSS Custom Properties is design tokens. Naming tokens is crucial and delicate—well-named tokens ensure consistency, ease of understanding, and scalability. When naming design tokens.

Reviewer 3: @dmtrmrv I think it would be worth while mentioning here to avoid the "nesting" of CSS variables, some projects map a color name to a ui name to another variable and you land up with 3 tokens for every one color - abstracting away 3 levels deep. We should try keep it somewhat high level so that the browser doesnt land up loading 200 tokens of which we only referencing 50 in the code base itself.

Reviewer 2: I may reference the use and value of primitive vs semantic vs contextual tokens (or custom properties of json tokens are not being used) .. or tier 1, 2, 3 (Brad Frost) .. but perhaps as a more general reference to start and something we can expand on in phase 2. In this case, the raw or default values are the primitives where the semantic and contextual tokens reference the primitives. 

Although, we are talking about naming in this context - so we don't want to go too far off topic.

Author: Agree, updated. Kept it high-level because the focus here on CSS and not design tokens.

* Use clear, meaningful names that reflect hierarchical relationships (e.g., `color-background-primary`, `spacing-medium`).  
* Avoid overly abstract, compact, or vague names; clarity and predictability are key. `button` is better `btn.`  
* Consistently apply naming patterns to help users anticipate token structures.  

> * Reference [Nathan Curtis's detailed guide](https://medium.com/eightshapes-llc/naming-tokens-in-design-systems-9e86c7444676#08b2) for in-depth practices.

Reviewer 2: YES - love this article!

Author: I come back to it once a week!

### Animations (@keyframes)

* Use clear action-based names (e.g., fade-in, slide-up).  

> * Maintain short, intuitive names indicating the animation’s behavior.

Reviewer 3: @dmtrmrv if using ITCSS, its worth noting where Animations should be placed within the triangle.

Author: I've added it in the ITCSS section.

### Grid Template Areas and Lines

* Area names should reflect their content purpose (e.g., header, sidebar, content).  
* Grid line names should be descriptive (e.g., main-start, footer-end).  
* Stick with concise, easily recognizable terms.

<h2 id="new-features" class="anchor-heading">New CSS Features {% include Util/link_anchor anchor="new-features" %}</h2>

> Browsers keep evolving, and CSS is always introducing new features. But adopting them too soon can cause issues for older or niche browsers. Here’s how we  evaluate and adopt modern CSS.

Reviewer: ```suggestion
Browsers keep evolving, and CSS is always introducing new features. But adopting them too soon can cause issues for older or niche browsers. Here’s how we evaluate and adopt modern CSS.
```

* Use [Can I Use](https://caniuse.com/) (backed by Google Baseline) to see which browsers support a feature.  
* Google Baseline classifies support:  
  * **Newly available:** Interoperable across all major browsers.  
  * **Widely available:** Feature has been interoperable for at least 30 months.  
  * **Limited availability:** Not yet in Baseline; some browsers lack support.

### 2. Provide Fallbacks

Place a widely supported declaration first, then the modern one. Old browsers ignore unknown properties, so they naturally fall back.

```css
.element {
  height: 100vh;    /* widely supported unit */
  height: 100dvh;   /* newer unit */
}
```

If the browser doesn’t know `100dvh`, it sticks to `100vh`. No breakage, just a simpler layout.

### 3. Use @supports

`@supports` is a native CSS feature-detection mechanism.

```css
.site {
  min-height: 100vh;
}

@supports (min-height: 100dvh) {
  .site {
    min-height: 100dvh;
  }
}
```

Or reverse it with not:

```css
.site {
  min-height: 100dvh;
}

@supports not (min-height: 100dvh) {
  .site {
    min-height: 100vh;
  }
}
```

> This "not" approach is easy to remove when older browsers catch up. Just look for a comment like `// remove` when `dvh` is fully supported.

Reviewer: `//remove` is not valid CSS comment

```suggestion
This "not" approach is easy to remove when older browsers catch up. Just look for a comment like `/* remove */` when `dvh` is fully supported.
```

### 4. Consider Polyfills

* Polyfills bring modern features to older browsers via JS.  
* They add dependencies, increase bundle size, and require maintenance.  
* If the feature can degrade gracefully, skipping a polyfill is often better.

### 5. Use Automated Tooling

* PostCSS (with postcss-preset-env) can convert cutting-edge CSS into more compatible code.  

> * Inspect how it transforms your code. Some transformations, like `@layer`, can bloat your CSS with hacks like `:not(#\#)` selectors.  

Reviewer 3: @dmtrmrv I think we want to be very specific about not using `@layer` in vanilla CSS implementations - the PostCSS transformation for this is awful

<h2 id="document-system" class="anchor-heading">Establish and Document the System {% include Util/link_anchor anchor="document-system" %}</h2>

When you start a new codebase, pause before writing components and choose a CSS methodology—such as ITCSS, BEM, or a utility-first approach—that fits the project’s size, maintenance horizon, and team expertise. Committing to a structured methodology early establishes clear naming rules, predictable specificity, and sensible file organization. These guardrails ease onboarding, minimize refactors, and save both time and budget over the life of the project.

If you opt to design a custom architecture, capture every key choice: folder layout, naming conventions, tooling, and any deviations from common patterns. A short explanation of why each decision was made helps future contributors understand the system quickly and avoids guesswork.

> **Note:** Keep this record where engineers will look first—typically in a /decisions/ directory inside the repository's setup or onboarding documentation. Follow your organization's standard documentation format so the notes remain consistent, searchable, and easy to maintain.

Reviewer 4: suggestion: Replace "Follow your organizations's" with "Follow 10up's" or "Follow your Project's"

Author: Agree, updated.

When building components, always start with the mindset that you have no control over the content they’ll handle. Even if the client or designer specifies certain limits, assume those constraints could change—or be ignored entirely.

Content is unpredictable. You might not know how long a title will be, how many items a navigation menu will have, or the length of an excerpt in a card. Design decisions made today can quickly be overturned tomorrow when new requirements come in.

Prepare for edge cases. Your components must be resilient and robust enough to handle all reasonable variations, including unexpected ones. If you don’t account for these possibilities, it could lead to broken layouts, unreadable text, or poor user experiences.

**Useful resources:**

- [Defensive CSS](https://defensivecss.dev/)

### Stress Test Your Components

When building a component, test it against extreme content scenarios. For example:

* A navigation menu with 20 items instead of 5.  
* A card title that wraps onto three or more lines.  
* A card title that is just one word.  

> * An excerpt that's 300 characters long—or empty.

Reviewer 4: suggestion: "A card without a thumbnail" <-- this I see happening more often than not.

Author: Def agree, updated.

**Useful resources:**

- [The Inverted Triangle Architecture: how to manage large CSS Projects](https://www.freecodecamp.org/news/managing-large-s-css-projects-using-the-inverted-triangle-architecture-3c03e4b1e6df/)
- [ITCSS × Skillshare](https://csswizardry.com/2018/11/itcss-and-skillshare/) - Harry Roberts' official collaboration on ITCSS methodology

<h2 id="accessibility" class="anchor-heading">Always Keep Accessibility in Mind {% include Util/link_anchor anchor="accessibility" %}</h2>

### Keyboard Navigation

Make sure all interactive elements are accessible via the keyboard. Test that users can navigate through the interface using the Tab key and activate actions with Enter or Space. Avoid creating elements that require a mouse or touch to work.

### Color Contrast

Ensure there’s enough contrast between text and its background. Use tools to check for WCAG compliance. Aim for a contrast ratio of at least 4.5:1 for normal text and 3:1 for larger text. Avoid relying on color alone to convey information.

### Focus

> Properly manage focus states. Use :focus styles to make focused elements visible and distinct. Don’t trap focus in modal dialogs or dropdowns without providing a clear way to exit. Reset focus logically when users interact with dynamic content, such as moving to a new page.

Reviewer 5: @dmtrmrv I tend to use :focus-visible more nowadays, do you have an opinion?

  margin-block: 0;
}
```

**Useful resources:**

- [`:where()` on MDN](https://developer.mozilla.org/en-US/docs/Web/CSS/:where) - Comprehensive documentation on specificity impacts and browser support

### Avoid Excessive Nesting

Limit the depth of nesting in your CSS selectors. Two levels of nesting are acceptable, and three should only be used when absolutely necessary. Excessive nesting can make CSS harder to read and maintain. It also increases specificity unnecessarily, which complicates overriding styles and debugging. Over-nesting complicates maintenance and increases specificity unnecessarily.

It’s okay to nest pseudo classes that represent state, for example:

```css
/* block */
.button {

  /* ✅ state inside the block is fine */
  &:hover {
    /* styles here */
  }
}

/* ✅ element and modifier stay flat */
.button__icon {
  /* styles here */
}

.button--primary {
  /* styles here */
}
```

It's not necessary to nest elements or modifiers:

```css
/* block */
.button {

  /* state inside the block is fine */
  &:hover {
    /* styles here */
  }

  /* ❌ element and modifier must not be nested inside the block */
  & .button__icon {
    /* styles here */
  }

>   & .button--primary {

Reviewer 5: @dmtrmrv i think the example has a typo? super nitpick, if the modifier is nested it would have been `&.button--primary` ?

Author: Def a typo, see I don't even know how to nest things...
